concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState/Modules.hs

```diff
@@ -398,19 +398,69 @@ migrateModules ::
 
-migrateModule :: 
-  (SupportsPersistentModule m, SupportsPersistentModule (t m), SupportMigration m t)
+migrateCachedModule ::
+  forall t m . (SupportsPersistentModule m, SupportsPersistentModule (t m), SupportMigration m t)
   => CachedModule
   -> t m CachedModule
-migrateModule mdl = do
-    newModule <- lift (refLoad mdl)
-    ref <- refMake newModule
-    (ret, _) <- refFlush ref
-    return ret
+migrateCachedModule cm = do
+  existingModule <- lift (refLoad cm)
+  case existingModule of
+    ModuleV0 v0 -> migrateModuleV v0
+    ModuleV1 v1 -> migrateModuleV v1
+  where
+    migrateModuleV ::
+      forall v .
+      (IsWasmVersion v) =>
+      ModuleV v
+      -> t m CachedModule
+    migrateModuleV mv@ModuleV{..} = do
+      newModuleVSource <- do
+        s <- lift (loadRef moduleVSource)
+        storeRef s
+      newModule <- sudV (getWasmVersion @v) newModuleVSource mv
+      let h :: Hash = getHash newModule
+      (flushedModule, _) <- refFlush =<< makeHashedCachedRef newModule h
+      return flushedModule
+    sudV :: SWasmVersion v -> BlobRef (WasmModuleV v) -> ModuleV v -> t m Module
+    sudV ver newSource ModuleV{moduleVInterface = GSWasm.ModuleInterface{..}} = do
+            !instrumentedModuleBytes <- case miModule of
+              PIMVMem instrModule -> return $ case ver of
+                  SV0 -> encode instrModule
+                  SV1 -> encode instrModule
+              PIMVPtr ptr -> do
+                artifact <- lift (loadBlobPtr ptr)
+                return $ runPut $ do
+                  -- This must match the serialization of InstrumentedModuleV
+                  put (demoteWasmVersion ver)
+                  putWord32be (fromIntegral (BS.length artifact))
+                  putByteString artifact
+            let headerBytes = runPut $ do
+                  put miModuleRef
+                  putSafeSetOf put miExposedInit
+                  putSafeMapOf put (putSafeSetOf put) miExposedReceive
+                footerBytes = runPut $ do
+                  putWord64be miModuleSize
+                  put newSource
+                !headerLen = fromIntegral $ BS.length headerBytes
+                !imLen = fromIntegral $ BS.length instrumentedModuleBytes
+            br <- storeRaw (headerBytes <> instrumentedModuleBytes <> footerBytes)
+            let !miModule' = PIMVPtr BlobPtr {
+                    -- Pointer is blob ref + 8 bytes (length of blob) + header length +
+                    -- 4 bytes for version + 4 bytes for length of instrumented module
+                    theBlobPtr = theBlobRef br + 8 + headerLen + 8,
+                    -- Length is the length of the serialized instrumented module -
+                    -- 4 bytes for version - 4 bytes for length
+                    blobPtrLen =  imLen - 8
+                  }
+            let mv' = ModuleV{moduleVInterface = GSWasm.ModuleInterface{miModule = miModule',..}, moduleVSource = newSource,..}
+            return $! mkModule ver mv'
```

Comment:
Both `sudV` and `refFlush` will end up writing the module to the blob store, which is redundant.  Perhaps something like the following would do:

```suggestion
      newModuleVInterface <- lift $ mapM (fmap PIMVMem . loadInstrumentedModuleV) moduleVInterface
      (newModuleBlobRef, newModule) <- storeUpdateDirect $
          mkModule (getWasmVersion @v) $ ModuleV {
            moduleVInterface = newModuleVInterface,
            moduleVSource = newModuleVSource
          }
      let !h :: Hash = getHash newModule
      return $! HCRFlushed newModuleBlobRef h
```

For a cleaner abstraction, the `storeUpdateDirect`, `getHash(M)` and `HCRFlushed` could be moved to a `makeFlushedHashedCachedRef`.

Reply:
Yes, I realized that. I do not like deserializing the artifact which is what `loadInstrumentedModuleV` does, so there is a bit of duplication with `storeDirect/loadDirect` in the current version.

but I added documentation to all these functions so that they will hopefully be updated at the same time.

Reply:
`loadInstrumentedModuleV` no longer deserializes the artifact.

Reply:
It does on the branch that this PR is based on, i.e., main.

Reply:
```
-- |Load an instrumented module from a 'BS.ByteString'. This creates a reference counted copy of
-- the artifact on the Rust side and attaches a finalizer to drop the reference when it is
-- garbage collected on the Haskell side.
instrumentedModuleVFromBytes :: forall v. (IsWasmVersion v) => BS.ByteString -> IO (InstrumentedModuleV v)
instrumentedModuleVFromBytes bs = do
    artifact <- moduleArtifactFromBytes @v bs
    case getWasmVersion @v of
      SV0 -> return (InstrumentedWasmModuleV0 artifact)
      SV1 -> return (InstrumentedWasmModuleV1 artifact)
```